Type-bind for submission input (port from DSpace-CRIS 7)

kshepherd opened the pull request:
## References
_Add references/links to any related issues or PRs. These may include:_
* Fixes #2873 
* Related to [REST Contract - submissionforms.md](https://github.com/DSpace/RestContract/blob/main/submissionforms.md)
* Related to https://github.com/DSpace/dspace-angular/pull/1529 (dspace-angular PR)

## Description
A port of type-bind functionality ported from DSpace CRIS 7.

The type-bind XML element is re-implemented in submission-forms.dtd and submission-forms.xml and the value, if any, is included in the submission form config REST resource.

In addition to the port, the field to check for item type has been made configurable, and integration tests have been written.

Original work created by 4Science for DSpace CRIS 7. Ported and modified by The Library Code GmbH with support from Technische Universität Berlin.

## Instructions for Reviewers

### Changes
* submission-forms xml and dtd are updated to accept an optional type-bind element with comma-seperated values eg. `<type-bind>Book, Article</type-bind>`. This matches previous functionality prior to DSpace 7.
* **DCInput** is updated with a new getter to return the full list of type bind values
* **SubmissionFormFieldRest** is updated with a typeBind field, getter and setter (`List<String>`)
* **SubmissionFormConverter** is updated to set the form field typeBind list from the DCInput if it is a metadata field
* **MetadataValidation** is updated to strip metadata values from the input data (and the existing workspaceitem, if relevant) for non-allowed fields (for current item type)
* **dspace.cfg** is updated to include a `submit.type-bind.field` property (default: dc.type)

### Tests
* **SubmissionFormFieldMatcher** is updated to include typeBind in the field matcher
* **SubmissionFormsControllerIT** is updated to bind the dc.relation.ispartofseries input field to the Technical Report type, and a new integration test written to expect this binding when reading the mock traditionalpageone object
* **WorkspaceItemRestRepositoryIT** is updated to test patch add operations for fields with type binding

To test, try configuring type-bind in submission-forms as you would have in DSpace 6 or earlier. The describe forms should update to reflect which fields are allowed for the current item type dynamically. (defaults to dc.type which is the usual "Type" input dropdown in traditional publication forms). The results can be reviewed in the REST HAL Browser for the submission form being tested, and/or combined with an end-to-end test in dspace-angular (see related PR: )

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] My PR does NOT include any new, third-party dependencies
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

kshepherd committed:
[TLC-254] Rebase main (merges to MetadataValidation)

kshepherd committed:
[TLC-254] Rebase to main (merge metadatavalidation)

kshepherd committed:
[TLC-254] Make item type field configurable for type bind (default dc.type)

kshepherd committed:
[TLC-254] Integration test (including changing matcher methods) for type bind

tdonohue commented:
@kshepherd : Wanted to make sure you are aware this PR is currently failing Integration Tests.  It'd be nice to get those fixed, as that makes reviews easier.

artlowel commented:
Something I noticed during my review of the angular PR:

If you select a certain type, fill out fields specific to it.  Then select another (so those fields disappear) and submit the item. The values you originally filled out for the wrong type will still end up in the Item. I'm unsure if it worked this way in DSpace 6, just thought I'd mention it in case it's a bug

kshepherd committed:
[TLC-254] Rebase main (merges to MetadataValidation)

kshepherd committed:
[TLC-254] Fix typeBind JSON path test

kshepherd committed:
[TLC-254] Fix typeBind JSON path test and add opposite test

kshepherd committed:
[TLC-254] If type bind doesn't allow a field, remove from data

kshepherd committed:
[TLC-254] Simplify validation test skip

kshepherd committed:
[TLC-254] Add ITs to ensure values for non-bound types don't persist

kshepherd committed:
[TLC-254] Remove old type-bound values from workspace items at deposit

This is also done at DescrbeStep getData() but if an item was saved
while it had a valid type for the given field, the validation method
also needs to remove it from the workspace item.

kshepherd committed:
[TLC-254] fix accidental codestyle violation in new import

kshepherd committed:
[TLC-254] fix accidental codestyle violation in new import

kshepherd committed:
[TLC-254] Use of context for just narrow type bind cases

kshepherd committed:
[TLC-254] Update type bind ITs to reflect validation changes

kshepherd committed:
[TLC-254] Obtain context properly in MetadataValidation

kshepherd commented:
@tdonohue This PR is ready for review, please let me know if there are any additional questions I can help with

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/DescribeStep.java:
```diff
@@ -45,6 +46,10 @@ public class DescribeStep extends AbstractProcessingStep {
 
     private DCInputsReader inputReader;
 
+    private static final String DOCUMENT_TYPE_FIELD =
```
Could we move this out of a static global field?  Otherwise, I believe this value will be cached until you next restart Tomcat (which means this config won't reload automatically).  Instead, I'd recommend just loading this config where you need it in the methods below -- there's only one usage, so it should be a quick fix.

kshepherd replied:
Done!

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/MetadataValidation.java:
```diff
@@ -42,6 +44,10 @@ public class MetadataValidation extends AbstractValidation {
 
     private static final String ERROR_VALIDATION_REGEX = "error.validation.regex";
 
+    private static final String DOCUMENT_TYPE_FIELD =
```
Same here, let's move this out of static field and just call the ConfigurationService where you need the value in methods below.

kshepherd replied:
Done!

tdonohue commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/WorkspaceItemRestRepositoryIT.java:
```diff
@@ -1935,6 +1935,141 @@ public void patchUpdateMetadataTest() throws Exception {
+                .andExpect(status().isOk())
+                .andExpect(jsonPath("$.errors").doesNotExist())
+                .andExpect(jsonPath("$",
+                        // Check this - we should match an item with no series or type
```
Tiny thing...comments seem wrong here (next line is matching an item with the expected series & type).

Same thing on lines 2037, and 2043 below. The comments are incorrect, but the code is correct.

tdonohue commented on dspace/config/dspace.cfg:
```diff
@@ -920,6 +920,11 @@ metadata.hide.dc.description.provenance = true
 
+# Which field should be used for type-bind
+# Defaults to 'dc.type'; If changing this value, you must also update the related
+# dspace-angular environment configuration property submission.typeBind.field
```
Small thing... we don't need the same configuration on *both* the UI and the backend.  Instead, the UI can read this config **from** the backend, if we add this new config to `rest.properties.exposed` in `rest.cfg`.  I.e...
```
rest.properties.exposed = submit.type-bind.field
```

With that field set, this property will be accessible via the REST API at
`[dspace.server.url]/api/config/properties/submit.type-bind.field`

This would allow you to completely **remove** the `typeBind` config from the Angular side.

kshepherd replied:
This should be done

tdonohue reviewed (changes requested):
@kshepherd : Overall, this code looks great & it works as described.  However, I have a few minor suggestions inline below. I think these should all be quick to cleanup.

kshepherd committed:
[TLC-254] Expose type bind field config with REST

kshepherd committed:
[TLC-254] Make type bind config dynamic, fix test comments

tdonohue reviewed (approved):
👍 Thanks @kshepherd !  Retested today and this backend PR looks to be working as described in the REST Contract.  Thanks for addressing all my feedback.

tdonohue commented:
Merging with `1 Approval` as this is ported code and a small PR (over 1/2 this PR is tests).  Thanks again @kshepherd !
